### DISCUSSION TITLE
serde_as on struct or struct tuple values

### DISCUSSION THREAD

**Discussion by leighmcculloch** [jonasbb/serde_with#518] [Q&A] [answered]
Hi! I have a type defined as follows, and it currently serializes/deserializes to an array of numbers.

I'm hoping to make the type always serialize as base64 anywhere it is used.

```rust
#[derive(serde::Serialize, serde::Deserialize)]
pub struct VecM<T>(Vec<T>);
```

All the examples I've found of using `serde_as` show using it on a field, not on a struct.

When I attempt to use it on a struct like as follows I get the following error:

```rust
#[derive(serde::Serialize, serde::Deserialize)]
#[serde_as(as = "Base64")]
pub struct VecM<T>(Vec<T>);
```

```
error: Unknown field: `as`
   --> src/next.rs:569:12
    |
569 | #[serde_as(as = "Base64")]
```

When I attempt to use it on the struct field like as follows I get the following error:

```rust
#[derive(serde::Serialize, serde::Deserialize)]
pub struct VecM<T>(
    #[serde_as(as = "Base64")]
    Vec<T>
);
```

```
error: expected non-macro attribute, found attribute macro `serde_as`
   --> src/next.rs:570:3
    |
570 | #[serde_as(as = "Base64")]
    |   ^^^^^^^^ not a non-macro attribute
```

Is there a way to attach the base64 serializer to a type rather than at each location it is used in a field?

**Comment by jonasbb** [comment, marked as answer]
The `serde_as` attribute is a replacement for serde's with attribute. As such, it only works on fields and not on structs. https://docs.rs/serde_with/2.0.1/serde_with/guide/serde_as/index.html#switching-from-serdes-with-to-serde_as

Your last example looks almost right. You need an additional attribute before the derive. Try adding `#[serde_with::serde_as]` at the beginning.

If your `Base64` type is also from `serde_with` there is another problem. The `Base64` type only works with types implementing `AsRef<[u8]>`, but `Vec<T>` is not one of them. It only works for the special case when your `T` is `u8`.
